presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -71,11 +71,22 @@ public PushPartialAggregationThroughExchange(FunctionRegistry functionRegistry)
     @Override
     public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
     {
+        PlanNode childNode = context.getLookup().resolve(aggregationNode.getSource());
+        if (!(childNode instanceof ExchangeNode)) {
+            return Result.empty();
+        }
+
+        ExchangeNode exchangeNode = (ExchangeNode) childNode;
+
         boolean decomposable = aggregationNode.isDecomposable(functionRegistry);
 
         if (aggregationNode.getStep().equals(SINGLE) &&
                 aggregationNode.hasEmptyGroupingSet() &&
-                aggregationNode.hasNonEmptyGroupingSet()) {
+                aggregationNode.hasNonEmptyGroupingSet() &&
+                exchangeNode.getType() == REPARTITION) {
```

Comment:
I think this condition won't fire at all, because `AddLocalExchanges` right now will always add gathering local exchange below final aggregation. This is fixed via https://github.com/prestodb/presto/pull/10205

Reply:
You're right. I'll rebase on top of that PR.